### ISSUE TITLE
Auto indent on paste behaves oddly with already indented lines.

### ISSUE THREAD

**Issue by NatalieWolfe** [atom/atom#5473]
When copying a segment of code that starts with a new-line followed by tabbed-in code while "Auto Indent On Paste" is enabled, consecutive pastes will each be an additional tab deeper. I tested this with C++, Java, and JavaScript which makes me believe it is not a problem with the language specs.

Here is an example in the form of an animated gif:
![autoindent-2](https://cloud.githubusercontent.com/assets/466761/6118758/d4ce2e04-b075-11e4-864b-46726fcd809e.gif)

It isn't just comments that this happens with, but it does require the previous line to be selected. Specifically, this happens when the copy-buffer matches a regex like `/^\n\s+[^\s]/`. That is, it starts with a newline, followed by one or more whitespace characters and then a non-whitespace character. And the previous line must be a blank line.

**Comment by 50Wliu** [comment]
What version of Atom is this on?

/cc @maxbrunsfeld

**Comment by NatalieWolfe** [comment]
Oops, should have included that. I'm using Mac v0.177.0, but I've been experiencing this bug for a few versions now (not entirely sure when it first presented).

Interesting, just tried to reproduce this on my Windows build box too. It has v0.175.0 installed. On that machine the behavior is reversed. When Auto Indent On Paste is enabled then pasting behaves as expected (each consecutive line is pasted at the same depth). However when that option is disabled then the first paste behaves as expected, but the 2nd through nth pastes are all one level deeper than the last. Not sure how to update the Windows version to see if it persists in 0.177.0.

**Comment by izuzak** [comment]
@NatalieWolfe I'm having trouble reproducing this in the latest version of Atom:

```
$ atom -v
0.180.0-2dac7fd
```

Here's what I'm doing:

![paste](https://cloud.githubusercontent.com/assets/38924/6190840/29790b7c-b3a5-11e4-90b2-9a12c08adc91.gif)

Is there something I'm doing differently than you are?

Can you try reproducing this again and record a gif in which the invisible characters are shown as well as the keybinding resolver (`cmd+.`)? Also, can you record the gif so that includes the point at which you select and copy text into the clipboard for the first time (it's not shown in the gif above)?

Atom should update to the latest version on it's own, but you can download the latest version here as well:

https://github.com/atom/atom/releases/latest

**Comment by NatalieWolfe** [comment]
@izuzak: The link you provided goes to 0.179.0, which is what I have installed on my Mac at the moment. 

Here is an updated gif, starting from a completely new window. Auto Indent on Paste is enabled during this demonstration.

![autoindent-3](https://cloud.githubusercontent.com/assets/466761/6192519/c1f50dd0-b367-11e4-9aec-92fe1f880da7.gif)

If you would like me to make another where I type slower (so you have more time to see the keybinding resolver updates) let me know.

**Comment by NatalieWolfe** [comment]
I notice on your gif, @izuzak, that you are using a plain text file. I don't think Auto Indent on Paste works in plaintext files (there is no structure to the text for it to use).

**Comment by izuzak** [comment]
> I notice on your gif, @izuzak, that you are using a plain text file. I don't think Auto Indent on Paste works in plaintext files (there is no structure to the text for it to use).

Ahhhh, that was it, thanks @NatalieWolfe! Seeing it now as well.

**Comment by karai17** [comment]
I am running the current Atom (194) on Windows 7 and this issue persists.

**Comment by muppetjones** [comment]
There are many indentation problems, and I'm trying to link all of them together in a single issue: [here](https://github.com/atom/atom/issues/6655)

**Comment by jeancroy** [comment]
This is what I believe I see here:

A) auto indent correctly identify the score indentation.
B) indent of the pasted text get added to baseline, creating wrong indentation.

If B is already done, there's another source of the problem:

C) indent of the  pasted text should be determined from a non empty line.

Because in this example first line was "\n" so a procedure like B cannot infer correct indentation.
A line with only withe-space is probably not so good either.

**Comment by JesseLeite** [comment]
I'm finding that `Auto Indent` works differently than `Auto Indent On Paste` setting in Atom.  There's a consistency issue between the two...

If I take this PHP method:
```php
        public function name()
    {
if ($something) {
    return $something;
}
            }
```

And paste it into a PHP class, I get:
```php
    public function name()
{
if ($something) {
return $something;
}
        }
```

So `Auto Indent On Paste` is doing something, but it's not doing it correctly.  However, if I now run `Auto Indent` from the Edit > Lines menu, I get this:
```php
    public function name()
    {
        if ($something) {
            return $something;
        }
    }
```
Beautiful, now it's properly intended for my PHP class.  Auto indentation on paste is something PHPStorm does without flaw (see http://g.recordit.co/Rj6unPomlW.gif).  It's weird that Atom can manually `Auto Indent` fine, but `Auto Indent On Paste` seems useless sometimes.

**Comment by JesseLeite** [comment]
...Oh I see what's going on here.  Let me follow up on my last comment.

`Auto Indent On Paste` is not fixing indentation within the code being pasted... it's just fixing it's indentation as a whole relative to parent code surrounding it:
![example1](http://g.recordit.co/o3a1KpSNdP.gif)

Whereas PHPStorm is also fixing indentation within the block:
![example2](http://g.recordit.co/jQiB5z5ApQ.gif)

IMHO, maybe there should be more consistency between manually running `Auto Indent`, and `Auto Indent On Paste`, since running `Auto Indent` manually is smart enough to fix the mess.

Thoughts?

**Comment by maxbrunsfeld** [comment]
@JesseLeite we used to do it the way you're suggesting: auto indent every line of the pasted text, but for many languages, our auto-indent is not smart enough to do it correctly, and people found it surprising that they code that they pasted was being altered like that.

**Comment by karai17** [comment]
Would it be possible to read indenting rules from the current grammar?


---
Landon Manning
lmanning17@gmail.com

**Comment by jeancroy** [comment]
One thing for sure is that first line is problematic. If I had to copy JesseLeite `public function(){` i'd probably start the selection at `p` up to the closing `}` thus loosing indent information. Empty line as shown in first example is also a case where first line does not contain proper indent information.

I'd however agree that full blown aot indent is probably not what we want on paste, because we have no garantee user will only copy/paste properly matched block scope.

**Comment by JesseLeite** [comment]
@maxbrunsfeld I see what you are saying. I would agree, that definitely is an issue.

However, based on terminology, it's a very confusing inconsistency for the end user to understand:
`Paste` + `Auto Indent` != `Auto Indent On Paste`

See what I'm getting at?

**Comment by timkelty** [comment]
Related comment: https://github.com/atom/atom/issues/4272#issuecomment-155822920

I think the desired behavior matches Sublime's `paste_and_indent`.
With that setting, the content's leading whitespace is removed when pasting, which prevents the indentation cascade.

**Comment by brentshermana** [comment]
I've created a solution to the problem [with this pull request](https://github.com/atom/atom/pull/11267). To paraphrase, when text is copied starting with a newline character, a metadata member 'indentBasis' can be incorrect. More detail if you follow the link.

My pull request needs testing on Windows! Can you guys help me out?

**Comment by TimvdLippe** [comment]
A new PR is up with a fix: #12793. This bug has been bugging me for a long time now and would love to get it fixed.

**Comment by patstew** [comment]
I've had another go at fixing this - see this package: https://atom.io/packages/fix-indent-on-paste
It would be useful to get feedback on whether that package works as people expect - if so I'll update PR #12793 with the same code

**Comment by larslue** [comment]
Thanks @patstew. I had the feeling that this bug still persists even after a full year of development. Your package fixes atom which is now working as expected on my machine. Thanks!

**Comment by markiragoldberg** [comment]
Thanks @patstew, your work made Atom much more usable for me.

**Comment by clusteratlas** [comment]
This is fucking awesome, thank you very much.

**Comment by CodeKiwi** [comment]
@patstew Awesome, this has been bugging me for a while. I did notice one unexpected thing. When pasting over the existing block it removes the indent when it shouldn't. Not that this is a particularly realistic use case (why would you paste over a block with the same content). Just thought you might like to know. See screenshots (first is copy, second is paste). 

<img width="392" alt="screen shot 2017-09-06 at 7 49 42 am" src="https://user-images.githubusercontent.com/592021/30089617-2d460934-92d8-11e7-8fe0-879321374d3a.png">
<img width="379" alt="screen shot 2017-09-06 at 7 49 50 am" src="https://user-images.githubusercontent.com/592021/30089616-2d436756-92d8-11e7-9051-2776c7290eec.png">

**Comment by lennartq** [comment]
@CodeKiwi  I actually duplicate a selection by Ctrl-C, Ctrl-V, Ctrl-V, pasting over the original. Not sure how common that is.

**Comment by TimvdLippe** [comment]
This might be fixed with #15976.

**Comment by Krinkle** [comment]
Still an issue.
